internal/dynamic/plugin/subprocess/subprocess.go

```diff
@@ -0,0 +1,211 @@
+	}
+}
+
+type subprocess struct {
+	cmdArgs    []string
+	env        map[string]string
+	stdoutHook func(line string)
+	stderrHook func(line string)
+	logger     *service.Logger
+
+	cmd    *exec.Cmd
+	mu     sync.Mutex
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+// New creates a new SubProcess instance.
+func New(
+	cmd []string,
+	env map[string]string,
+	options ...Option,
+) (SubProcess, error) {
+	if len(cmd) == 0 {
+		return nil, errors.New("command cannot be empty")
+	}
+	s := &subprocess{
+		cmdArgs: cmd,
+		env:     env,
+		logger:  nil,
+	}
+	for _, option := range options {
+		option(s)
+	}
+	return s, nil
+}
+
+func (s *subprocess) Start() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.cmd != nil {
+		return ErrProcessAlreadyStarted
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	cmd := exec.CommandContext(ctx, s.cmdArgs[0], s.cmdArgs[1:]...)
+	if s.env != nil {
+		cmd.Env = os.Environ() // Inherit current environment
+		for k, v := range s.env {
+			cmd.Env = append(s.cmd.Env, fmt.Sprintf("%s=%s", k, v))
+		}
+	}
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		cancel()
+		return fmt.Errorf("failed to create stdout pipe: %w", err)
+	}
+	stderrPipe, err := cmd.StderrPipe()
+	if err != nil {
+		stdoutPipe.Close()
+		cancel()
+		return fmt.Errorf("failed to create stderr pipe: %w", err)
+	}
+	if err := cmd.Start(); err != nil {
+		stdoutPipe.Close()
+		stderrPipe.Close()
+		cancel()
+		return fmt.Errorf("failed to start command: %w", err)
+	}
+	s.logger.Debugf("Subprocess started with PID: %d", cmd.Process.Pid)
+	s.wg.Add(3) // For stdout, stderr, and process wait goroutines
+	go s.readOutput(stdoutPipe, false)
+	go s.readOutput(stderrPipe, true)
+	go func() {
+		defer s.wg.Done()
+		err := cmd.Wait()
+		s.logger.Debugf("Subprocess with PID %d exited with error: %v", cmd.Process.Pid, err)
+	}()
+	s.cmd = cmd
+	s.cancel = cancel
+	return nil
+}
+
+func (s *subprocess) readOutput(pipe io.Reader, isStderr bool) {
+	defer s.wg.Done()
+	src := map[bool]string{false: "stdout", true: "stderr"}[isStderr]
+	log := s.logger.With("source", src)
+	scanner := bufio.NewScanner(pipe)
+	hook := func(line string) {}
+	if !isStderr && s.stdoutHook != nil {
+		hook = s.stdoutHook
+	}
+	if isStderr && s.stderrHook != nil {
+		hook = s.stderrHook
+	}
+	for scanner.Scan() {
+		line := scanner.Text()
+		hook(line)
+		log.Infof("%s", line)
+	}
+	if err := scanner.Err(); err != nil && err != io.EOF {
+		log.Warnf("error reading from subprocess: %v", err)
+	}
+}
+
+func (s *subprocess) IsRunning() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.cmd == nil {
+		return false
+	}
+	if s.cmd.ProcessState != nil && s.cmd.ProcessState.Exited() {
+		return false
+	}
+	return true
+}
+
+func (s *subprocess) Close(ctx context.Context) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.cmd == nil || s.cancel == nil {
+		s.logger.Tracef("Close called on a subprocess that is not running or already closed.")
```

Comment:
Consider Warning level.

Reply:
This can happen in "normal" use, so will keep quiet.